docs/website/docs/developers/debugging/gpu-approaches-tips.md

```diff
@@ -0,0 +1,163 @@
+---
+icon: material/chip
+---
+
+# GPU debugging approaches and tips
+
+This page aims to provide general approaches and practical tips for debugging
+GPU compiler/runtime correctness/performance issues in IREE.
+
+GPUs fundamentally have similar architectures and software stacks.
+We target GPUs from various vendors using different GPU APIs, but they share
+quite a lot common infrastructure in IREE.
+So the approaches and tips here should be widely applicable.
+For the ones that are specific to a particular kind of problem/component/GPU,
+they are prefixed with proper tags to be clear.
+
+## General methodology
+
+The difficulties associated with debugging typically arise from **isolating the
+problematic component and pinpointing the culprit**.
+Once done, the solution typically derives naturally.
+
+There are many components in the IREE stack; hierarchically we can categorize
+them into either the compiler or runtime bucket.
+
+* For compilers, there are multiple layers from the top to the bottom--frontend
+  input importers, IREE flow/stream compilation, IREE host/device compilation,
+  GPU backend in LLVM proper or GPU driver compiler for SPIR-V.
+* For runtime, we have fewer layers--IREE HAL drivers, and GPU driver.
+
+Any of the above components/layers can have bugs.
+It's important to reduce the potential surface area to make the problem more
+tractable.
+
+Once we have a more isolated case, the general methodology to pinpoint the
+exact culprit is to **1) collect and inspect the symptoms, 2) form hypothesis
+and run experiments to prove/refute the hypothesis, and 3) iterate**.
+
+### .. with shortcuts
+
+The above procedure is for facing a large problem with no clue, for example,
+when bringing up a new model end-to-end via IREE.
+
+Though most of the time, we can leverage existing facilities to avoid going down
+the full top-down hiearchical debugging procedure.
+For example, for regression happening on an existing model, CI or `git bitsect`
+might tell us directly the culprit commit.
+
+### .. using tools
+
+For issues with strong signals like crashing, it's also easier to pinpoint the
+exact culprit with dedicated tools--we can leverage various
+[sanitizers](./sanitizers.md) or debuggers.
+
+## Isolating the problematic component
+
+If we are facing a large problem without a clear clue, we need to isolate the
+problematic compiler or runtime layer first, typically by comparing with a
+working solution:
+
+* `[correctness/performance]` Sanitize the environment first--did you recently
+  updated the GPU SDK or driver?
+  Are others able to reproduce the issue?
+  If not what SDK / driver versions they are using?
+  Is your machine drawing enough power when benchmarking?
+  Is your machine connected with a mointor (for Vulkan)?
+  How long since you last rebooted your machine? 👻
+  Asking these questions and making sure the environment is proper can save
+  you hours of debugging sometimes.
```

Comment:
Changed the first to a list. I'm not sure actually extra link breaks can enable multiple paragraphs? Maybe I'm not doing things correct though..

Reply:
Ah, indentation is 4 spaces, not 2.

```markdown
* `[correctness/performance]` Sanitize the environment first.
  Asking these questions and making sure the environment is proper can save
  you hours of debugging sometimes:
    * Did you recently updated the GPU SDK or driver?
    * Are others able to reproduce the issue?
    * If not what SDK / driver versions they are using?
    * Is your machine drawing enough power when benchmarking?
    * Is your machine connected with a mointor (e.g., for Vulkan)?
    * How long since you last rebooted your machine? 👻

* `[correctness/performance]` We have multiple GPU targets/drivers in
    IREE--LLVMGPU/CUDA, LLVMGPU/HIP, SPIR-V/Vulkan, SPIR-V/Metal.

    For the _same_ GPU, we typically have two paths to target, e.g., CUDA/HIP
    or Vulkan for NVIDIA/AMD GPUs, Metal or Vulkan for Apple GPUs.

    If one path is correct/performant, we can diff against it to try isolate
    the problem--the common/shared compiler/runtime code is likely okay; what
    differs between paths is likely problematic.

* `[correctness/performance][vulkan]` Vulkan supports different GPUs.
    Similarly, if one GPU gives correct/performant result, we diff against it
    to find clues.

    Even more code in compiler/runtime are shared here; what's problematic is
    likely different capabilities triggering different CodeGen pipelines so
    revealing bugs in a particular CodeGen pipeline.

    Or there are driver issues from a particular vendor.
```

![image](https://github.com/openxla/iree/assets/4010439/bdfe80d9-e322-4b88-8bec-f517b3b3c20d)

Reply:
Ha, okay, I see now. Thanks!